### DISCUSSION TITLE
Consider using pixiecore for the network booting

### DISCUSSION THREAD

**Discussion by kvaps** [ltsp/ltsp#223] [Q&A]
There is an amazing project - [pixiecore](https://github.com/danderson/netboot/tree/master/pixiecore) which is providing simplicity and everything need for the network booting on server side. It includes:

- TFTP-server
- HTTP-server
- DHCP-proxy

It can simple be runing inside docker container, and it provides an API for the configuration, We're using LTSP and starting readonly LTSP-servers in Kubernetes using the similar logic, so I just came up with an idea to implement support for LTSP.

As new LTSP19 is more modular, we can implement this support via special command, eg: `ltsp pixiecore` command will generate tree of directories and static files, which can be used by pixiecore as API for the network booting:
https://github.com/danderson/netboot/blob/master/pixiecore/README.api.md

**Comment by alkisg** [comment]
If you're willing to join the LTSP developers team and implement and test and support that, sure, it'll be most welcome.

Otherwise I think we could arrange an ltsp-plugins project where this would fit.

**Comment by kvaps** [comment]
I will,
I've just recorded my thoughts here 🙂

**Comment by kvaps** [comment]
HI, I have successful experiment with HTTP booting using [pixiecore](https://github.com/danderson/netboot/tree/master/pixiecore#debianubuntu), all what you need is: 

disable dhcp options and tftp service in dnsmasq, let there only dhcp-range:
```bash
cat > /etc/dnsmasq.d/ltsp-dnsmasq.conf <<EOT
port=0
dhcp-range=192.168.67.20,192.168.67.250,12h
EOT
systemctl restart dnsmasq
```
start pixiecore:
```bash
pixiecore boot --dhcp-no-bind \
  /srv/tftp/ltsp/x86_64/vmlinuz \
  /srv/tftp/ltsp/x86_64/initrd.img \
  /srv/tftp/ltsp/ltsp.img \
  --cmdline='root=/dev/nfs nfsroot=/srv/ltsp ltsp.image=images/x86_64.img loop.max_part=9'
```

**Comment by alkisg** [comment]
I would imagine that dnsmasq is more stable than pixiecore, but it wouldn't hurt to have alternatives.
If we can make `ltsp pixiecore` behave similarly to `ltsp dnsmasq`, supporting BIOS and UEFI clients, iPXE menus etc, no problem then.

For the HTTP part, see my last comment in the HTTP issue.

**Comment by kvaps** [comment]
OK, I was just playing a little with this and wrote [prototype of pixiecore api-server](https://github.com/kvaps/ltsp19/blob/pixiecore/ltsp/server/pixiecore) for ltsp on shell :)

I faced with issue that it can't use ID function in api-mode https://github.com/danderson/netboot/issues/103

So I'll switch back to http booting method development for ipxe+dnsmasq https://github.com/ltsp/ltsp/issues/75

**Comment by alkisg** [comment]
Since this needs no immediate action in the LTSP code base, I've moved it to the community issues.

**Comment by kvaps** [comment]
I think it does not require any special actions from LTSP side.

I already provided an example in https://github.com/ltsp/community/issues/76#issuecomment-557526786

After implementing https://github.com/ltsp/ltsp/pull/90 it also could be simple used like:
```bash
pixiecore boot --dhcp-no-bind \
  /srv/tftp/ltsp/x86_64/vmlinuz \
  /srv/tftp/ltsp/x86_64/initrd.img \
  /srv/tftp/ltsp/ltsp.img \
  --cmdline='ip=dhcp ltsp.image={{ ID "/srv/ltsp/images/x86_64.img" }} loop.max_part=9'
```

without any additional services running.
